src/Parser.hs

```diff
@@ -5,113 +5,112 @@ module Parser
 
 parserQuality :: Parser Quality
-parserQuality =
-       parser "^" QMajor
-  <||> parser "M" QMajor
-  <||> parser "-" QMinor
-  <||> parser "m" QMinor
-  <||> parser "dim" QDiminished
-  <||> parser "o" QDiminished
-  <||> parser "aug" QAugmented
-  <||> parser "+" QAugmented
+parserQuality = choice $ parseQualfromString <$> qualStrings
   where
-    parser :: String -> Quality -> Parser Quality
-    parser literal quality =
-      do
-        _ <- string literal
-        return quality
+    qualStrings = [ ("^",   QMajor)
+                  , ("M",   QMajor)
+                  , ("-",   QMinor)
+                  , ("m",   QMinor)
+                  , ("dim", QDiminished)
+                  , ("o",   QDiminished)
+                  , ("aug", QAugmented)
+                  , ("+",   QAugmented )
+                  ]
+
+    parseQualfromString :: (String, Quality) -> Parser Quality
+    parseQualfromString (qualName, quality) = string qualName >> return quality
 
 parserHighestNatural :: Parser HighestNatural
 parserHighestNatural =
   do
-    constructor <- parserMajor
+    constructor <- parseMajToConstructor
     digits <- many1 digit
-    return $ constructor (read digits)
+    return $ constructor $ read digits
   where
-    parserMajor :: Parser (Int -> HighestNatural)
-    parserMajor =
-      do
-        major <- optionMaybe $ string "Maj" <||> string "M" <||> string "^"
-        return $
-          case major of
-            Just _ -> HighestNatural Major
-            Nothing -> HighestNatural NonMajor
+    parseMajToConstructor :: Parser (Int -> HighestNatural)
+    parseMajToConstructor =
+      do major <- optionMaybe $ string "Maj" <||> string "M" <|> string "^"
+         return $ HighestNatural $
+           case major of
+             Just _  -> Major
+             Nothing -> NonMajor
 
 parserSus :: Parser Sus
 parserSus =
-  do
-    msus <- optionMaybe parserSusPresent
-    let sus = maybe NoSus Sus msus
-    return sus
+  do msus <- optionMaybe parserSusPresent
+     return $ maybe NoSus Sus msus
   where
     parserSusPresent :: Parser Int
     parserSusPresent =
-      do
-        _ <- string "sus"
-        number <- optionMaybe $ many1 digit
-        return $ maybe 2 read number
+      do string "sus"
+         number <- optionMaybe $ many1 digit
+         return $ maybe 2 read number
 
 parserExtension :: Parser Extension
 parserExtension =
-  do
-    sharpOrFlat <- parserSf
-    number <- many1 digit
-    return $ sharpOrFlat $ read number
+  do sharpOrFlat <- parserSf
+     number <- many1 digit
+     return $ sharpOrFlat $ read number
   where
     parserSf :: Parser (Int -> Extension)
     parserSf =
-      do
-        sf <- oneOf "b#"
-        return $ case sf of
-          'b' -> ExtFlat
-          '#' -> ExtSharp
-          _ -> error "Unhandled possiblity in parser"
+      do sf <- oneOf "b#"
+         return $ case sf of
+                    'b' -> ExtFlat
+                    '#' -> ExtSharp
```

Comment:
The reason this is OK is fairly subtle; the name `sf` will only be used _if the parsing action that binds it succeeds_. Thus, in our pattern match that follows, _we know that `sf` can only be either '#' or b'_. Thus there's no reason to include that catch-all.

For what it's worth, I believe it's situations such as this that justify GHC's default behavior of not warning about non-exhaustive pattern-matching; we have contextual information from the parsing monad that (dynamically) guarantees we won't hit that pattern match with anything but the characters we (hopefully) parsed.